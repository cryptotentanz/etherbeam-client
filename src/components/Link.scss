.my-ulink:hover {
  text-decoration: underline;

  &.my-disabled-fg {
    text-decoration: inherit;
  }
}

.my-link,
.my-button-link,
.my-ulink {
  display: inline-flex;
  align-items: center;
  border: 0px;
  background-color: inherit;

  &.my-active {
    font-weight: bold;
  }
}
