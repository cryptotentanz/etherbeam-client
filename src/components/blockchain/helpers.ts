import { faCoins, faFile, faWallet, IconDefinition } from '@fortawesome/free-solid-svg-icons'
import { BigNumber, ethers } from 'ethers'
import numeral from 'numeral'
import { AddressType, Token } from 'libraries/types'

const ETH_DECIMALS = 18
const MAX_DECIMALS = 13
const UNISWAP_SWAP_URL = 'https://app.uniswap.org/#/swap'

export interface AmountPrintOptions {
  forceDecimals?: boolean
  decimals?: number
}

const printAmount = (amount: string | null, decimals: number, options?: AmountPrintOptions | null): string => {
  const { forceDecimals, decimals: forcedDecimals } = options ?? {}
  const decimalsCount = Math.min(forcedDecimals || decimals, MAX_DECIMALS)
  const decimalFormat = forceDecimals ? '0'.repeat(decimalsCount) : `[${'0'.repeat(decimalsCount)}]`

  return amount ? numeral(amount).format(`0,0.${decimalFormat}`) : '?'
}

export type Unit = 'wei' | 'kwei' | 'mwei' | 'gwei' | 'ether'

export const printEtherAmount = (amount: BigNumber | null, options?: AmountPrintOptions | null, unit?: Unit): string =>
  printAmount(amount ? ethers.utils.formatUnits(amount, unit) : null, ETH_DECIMALS, options)

export const printTokenAmount = (
  amount: BigNumber | null,
  token: Token,
  options?: AmountPrintOptions | null
): string => {
  const { decimals } = token

  return printAmount(amount ? ethers.utils.formatUnits(amount, decimals) : null, decimals, options)
}

export const shortenAddress = (address: string): string => {
  const start = address.substring(0, 6)
  const end = address.substring(address.length - 4)

  return `${start}...${end}`
}

export const getAddressTypeIcon = (type: AddressType): IconDefinition | null => {
  switch (type) {
    case AddressType.Wallet:
      return faWallet
    case AddressType.Contract:
      return faFile
    case AddressType.Token:
      return faCoins
    default:
      return null
  }
}

export const getUniswapTradeUrl = (inputHash: string | null, outputHash: string | null): string => {
  return `${UNISWAP_SWAP_URL}?${inputHash ? `inputCurrency=${inputHash}&` : ''}${
    outputHash ? `outputCurrency=${outputHash}` : ''
  }`
}
