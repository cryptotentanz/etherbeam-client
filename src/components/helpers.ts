import { UserType } from 'libraries/types'
import { faUser, faUserShield, IconDefinition } from '@fortawesome/free-solid-svg-icons'

export const getUserTypeIcon = (type: UserType): IconDefinition => {
  switch (type) {
    case UserType.Administrator:
      return faUserShield
    default:
      return faUser
  }
}
