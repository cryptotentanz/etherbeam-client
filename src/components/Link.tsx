import React, { FC, ReactNode } from 'react'
import { useHistory } from 'react-router'
import { IconProp } from '@fortawesome/fontawesome-svg-core'
import Icon from './Icon'
import './Link.scss'

type ColorProp = 'dark' | 'light' | 'primary' | 'secondary' | 'positive' | 'negative'
type SizeProp = 'm' | 'l' | 'xl'
type LabelBreakpointProp = 'xs' | 's' | 'm' | 'l'

interface LinkProps {
  href: string
  external?: boolean
  label?: string
  icon?: IconProp
  description?: string
  color?: ColorProp
  size?: SizeProp
  button?: boolean
  noUnderline?: boolean
  labelBreakpoint?: LabelBreakpointProp
  active?: boolean
  disabled?: boolean
  className?: string
  children?: ReactNode
}

const Link: FC<LinkProps> = ({
  href,
  external,
  label,
  icon,
  description,
  color = 'secondary',
  size,
  button,
  noUnderline,
  labelBreakpoint,
  active,
  disabled,
  className,
  children,
}: LinkProps) => {
  const history = useHistory()

  const getClassName = (): string => {
    const classNames = ['my-clickable', className]
    if (button) {
      classNames.push(`my-link-button my-${color}-bg`)
    } else {
      classNames.push(noUnderline ? 'my-link' : 'my-ulink')
      classNames.push(`my-${color}-fg`)
    }
    if (size) classNames.push(`my-size-${size}`)
    if (disabled) classNames.push('my-disabled-fg')
    if (active) classNames.push('my-active')

    return classNames.join(' ')
  }

  const onKeyUp = (event: React.KeyboardEvent<HTMLAnchorElement>): void => {
    if (event.key == 'Enter') runAction()
  }

  const runAction = (): void => {
    if (!disabled) history.push(href, {})
  }

  return (
    <a
      title={description}
      className={getClassName()}
      target={external ? '_blank' : undefined}
      rel="noreferrer"
      href={external && !disabled ? href : undefined}
      onClick={!external ? runAction : undefined}
      onKeyUp={!external ? onKeyUp : undefined}
      tabIndex={!disabled ? 0 : undefined}
    >
      {children ? (
        children
      ) : icon ? (
        <Icon icon={icon} labelBreakpoint={labelBreakpoint} label={label} iconAlt={button} />
      ) : (
        <span>{label}</span>
      )}
    </a>
  )
}

export default Link
