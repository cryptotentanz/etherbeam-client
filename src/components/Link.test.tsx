import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import { clearMocks, mockFontAwesome } from 'tests/helpers'
import { faExclamationCircle } from '@fortawesome/free-solid-svg-icons'
import Link from './Link'

mockFontAwesome()

describe('<Link />', () => {
  const description = 'Description'
  const label = 'Label'
  const icon = faExclamationCircle
  const href = '/href'
  const className = 'my-class'

  let component: ShallowWrapper

  describe('Default', () => {
    beforeAll(() => {
      clearMocks()

      component = shallow(<Link href={href} label={label} />)
    })

    it('should render', () => {
      const a = component.find('a')
      expect(a.prop('href')).toBeUndefined()
      expect(a.prop('target')).toBeUndefined()
      expect(a.prop('onClick')).toBeDefined()
      expect(a.prop('onKeyUp')).toBeDefined()
      expect(a.prop('tabIndex')).toBe(0)
      expect(a.hasClass('my-ulink')).toBeTruthy()
      expect(a.hasClass('my-secondary-fg')).toBeTruthy()
    })

    it('should render icon', () => expect(component.find('span').at(0).prop('children')).toBe(label))

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Icon', () => {
    beforeAll(() => {
      component = shallow(<Link href={href} label={label} icon={icon} />)
    })

    it('should render icon', () => {
      const icon = component.find('Icon')
      expect(icon.prop('icon')).toBeDefined()
      expect(icon.prop('label')).toBe(label)
    })
  })

  describe('Custom', () => {
    beforeAll(() => {
      component = shallow(<Link href={href}>{label}</Link>)
    })

    it('should render items', () => expect(component.find('a').prop('children')).toContain(label))
  })

  describe('Description', () => {
    beforeAll(() => {
      clearMocks()

      component = shallow(<Link href={href} label={label} description={description} />)
    })

    it('should render', () => expect(component.find('a').prop('title')).toBe(description))
  })

  describe('External', () => {
    beforeAll(() => {
      clearMocks()

      component = shallow(<Link href={href} label={label} external />)
    })

    it('should render', () => {
      const a = component.find('a')
      expect(a.prop('href')).toBe(href)
      expect(a.prop('target')).toBe('_blank')
      expect(a.prop('onClick')).toBeUndefined()
      expect(a.prop('onKeyUp')).toBeUndefined()
    })
  })

  describe('No underline', () => {
    beforeAll(() => {
      clearMocks()

      component = shallow(<Link href={href} label={label} noUnderline />)
    })

    it('should render', () => expect(component.find('a').hasClass('my-link')).toBeTruthy())
  })

  describe('Active', () => {
    beforeAll(() => {
      clearMocks()

      component = shallow(<Link href={href} label={label} active />)
    })

    it('should render', () => expect(component.find('a').hasClass('my-active')).toBeTruthy())
  })

  describe('Disabled', () => {
    beforeAll(() => {
      clearMocks()

      component = shallow(<Link href={href} label={label} disabled />)
    })

    it('should render', () => {
      const a = component.find('a')
      expect(a.prop('tabIndex')).toBeUndefined()
      expect(a.hasClass('my-disabled-fg')).toBeTruthy()
    })
  })

  describe('Class', () => {
    beforeAll(() => {
      clearMocks()

      component = shallow(<Link href={href} label={label} className={className} />)
    })

    it('should render', () => expect(component.find('a').hasClass(className)).toBeTruthy())
  })
})
