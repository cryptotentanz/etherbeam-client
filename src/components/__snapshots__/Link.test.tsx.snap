// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> Default should match snapshot 1`] = `
<a
  class="my-clickable  my-ulink my-secondary-fg"
  rel="noreferrer"
  tabindex="0"
>
  <span>
    Label
  </span>
</a>
`;
